### ISSUE TITLE
io.EOF error opening raw objects

### ISSUE THREAD

**Issue by TheDahv** [speedata/gogit#8]
## Short Version

Go Version: 1.7
GoGit @ 1396c978ff9beb13261ff75b4a46db6abded251e

`readObjectFile` fails with an `io.EOF` error when reading the zlib file reader into the buffer. 
The attached [diff patch](https://github.com/speedata/gogit/files/472900/object-read-eof-error.txt) shows what changes I made to get my program running again.

I think there either needs to be proper handling of the `EOF` error, or perhaps some education on what I might have done wrong in my program. Details below...

## Background

I recently upgraded Go to 1.7 and tried re-running [a project](https://github.com/TheDahv/git-reviewer/tree/feat-gogit-implementation__with-system-calls) I've been working on. I've been slowly transitioning away from [git2go](https://github.com/libgit2/git2go) so I can get away from the cgo dependency.

At this part of the code, I'm really just trying to open the commit pointed to by the `master` ref--regardless of what branch I might currently be in--so I can compare its date with the current `HEAD`.

## Usage

My client code looks like this (ignore [`rg.maybeRunMany`](https://github.com/TheDahv/git-reviewer/blob/feat-gogit-implementation__with-system-calls/src/gitreviewers/helpers.go#L40). Its job is just to keep an internal error value and run the chain of functions until the first errors):

```go
	rg.maybeRunMany(
		func() {
			baseRef, rg.err = r.RepoGogit.LookupReference("refs/heads/master")
			if rg.err != nil {
				rg.msg = "unable to open master branch"
			}
		},
		func() {
			baseCommit, rg.err = r.RepoGogit.LookupCommit(baseRef.Oid)
			if rg.err != nil {
				rg.msg = "unable to open master commit"
			}
		},
		func() {
			headRef, rg.err = r.RepoGogit.LookupReference("HEAD")
			if rg.err != nil {
				rg.msg = "unable to open HEAD ref"
			}
		},
		func() {
			headCommit, rg.err = r.RepoGogit.LookupCommit(headRef.Oid)
			if rg.err != nil {
				rg.msg = "unable to open HEAD commit"
			}
		},
		func() {
			behind = headCommit.Committer.When.Before(baseCommit.Committer.When)
		},
	)
	if rg.err != nil {
		fmt.Println("Error in GoGit chain:", rg.msg)
		fmt.Println(rg.err)
	}
	return behind, rg.err
```

## Result

The output from running this is as follows:

```bash
$ go run main.go
Error in GoGit chain: unable to open master commit
EOF
```

I added some logging to my gogit installation to narrow down to this function:

```
$ go run main.go
[GoGit] Searching for OID object. Looking in path /Users/davidpierce/Development/go/src/github.com/thedahv/git-reviewer/.git
[GoGit] objPath /Users/davidpierce/Development/go/src/github.com/thedahv/git-reviewer/.git/objects/99/01bf79f808a8339b9820c08e209f5ec9649bda
[GoGit] os.Stat error false
[GoGit][readObjectFile] File open issue for /Users/davidpierce/Development/go/src/github.com/thedahv/git-reviewer/.git/objects/99/01bf79f808a8339b9820c08e209f5ec9649bda: false
[GoGit][readObjectFile] first_buffer_size: 1024
[GoGit][readObjectFile] First buffer read error
[GoGit][readObjectFile] Bytes read 271
[GoGit] Opening raw for OID 9901bf79f808a8339b9820c08e209f5ec9649bda
[GoGit] OID open error EOF
Error in GoGit chain: unable to open master commit
EOF
There was an error determining branch state: EOF
```

So what do you think? Did I break something or am I misunderstanding how this should work? Why have I not seen this before upgrading to Go 1.7?

Again, handling the `EOF` error in `readObjectFile` fixes it for me locally. If you examine the patch, you'll see the logging I added, and a simple naive check to look for an `EOF` error and ignore it when encountered.

**Comment by TheDahv** [comment]
FWIW, here's where I went to read up about the EOF error and how consumers of Readers are supposed to deal with them:

- [io.Reader](https://golang.org/pkg/io/#Reader)
- [io.EOF](https://golang.org/pkg/io/#pkg-variables)

There are other functions that return the contents of the file--though, perhaps not as efficiently as what you may be trying to do with your fixed-length buffer--that hide the `EOF` details away from the client code. I don't have as much familiarity with what other needs are met by this function, so I couldn't recommend either way.

**Comment by pgundlach** [comment]
I'll have a look, but unfortunately I won't be able to answer until the day after tomorrow.

**Comment by TheDahv** [comment]
Oh! I see you closed the issue. Is that to communicate that this project isn't supported and the fix won't be considered?

**Comment by pgundlach** [comment]
No, not at all. I thought the issue was fixed with the pull request (#9). I have included that pull request, so I am done (<-- at least, this is what I tought).

**Comment by TheDahv** [comment]
Oops, I misread and didn't see that. Ha, I didn't expect it to get merged so fast. 

Thank you :)
